src/gui/widgets/LinkedModelGroupViews.cpp

```diff
@@ -0,0 +1,159 @@
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "LinkedModelGroupViews.h"
+
+#include <QPushButton>
+#include "Controls.h"
+#include "ControlLayout.h"
+#include "LinkedModelGroups.h"
+
+
+/*
+	LinkedModelGroupViewBase
+*/
+
+
+LinkedModelGroupView::LinkedModelGroupView(QWidget* parent,
+	LinkedModelGroup *model, std::size_t colNum) :
+	QWidget(parent),
+	m_model(model),
+	m_colNum(colNum),
+	m_layout(new ControlLayout(this))
+{
+}
+
+
+
+
+LinkedModelGroupView::~LinkedModelGroupView() {}
+
+
+
+
+void LinkedModelGroupView::modelChanged(LinkedModelGroup *group)
+{
+	// reconnect models
+	group->foreach_model([this](const std::string& str,
+		const LinkedModelGroup::ModelInfo& minf)
+	{
+		auto itr = m_widgets.find(str);
+		// in case there are new or deleted widgets, the subclass has already
+		// modified m_widgets, so this will go into the else case
+		if (itr == m_widgets.end())
+		{
+			// no widget? this can happen when the whole view is being destroyed
+			// (for some strange reasons)
+		}
+		else
+		{
+			itr->second->setModel(minf.m_model);
+		}
+	});
+
+	m_model = group;
+}
+
+
+
+
+void LinkedModelGroupView::addControl(Control* ctrl, const std::string& id,
+	const std::string &display, bool removable)
+{
+	int wdgNum = static_cast<int>(m_widgets.size());
+	if (ctrl)
+	{
+		QWidget* box = new QWidget(this);
+		QHBoxLayout* boxLayout = new QHBoxLayout(box);
+		boxLayout->addWidget(ctrl->topWidget());
+
+		if (removable)
+		{
+			QPushButton* removeBtn = new QPushButton;
+			removeBtn->setIcon( embed::getIconPixmap( "discard" ) );
+			QObject::connect(removeBtn, &QPushButton::clicked,
+				this, [this,ctrl](bool){
+					AutomatableModel* controlModel = ctrl->model();
+					// remove control out of model group
+					// (will also remove it from the UI)
+					m_model->removeControl(controlModel);
+					// delete model (includes disconnecting all connections)
+					delete controlModel;
+				},
+				Qt::DirectConnection);
+			boxLayout->addWidget(removeBtn);
+		}
+
+		// required, so the Layout knows how to sort/filter widgets by string
+		box->setObjectName(QString::fromStdString(display));
+		m_layout->addWidget(box);
+
+		m_widgets.emplace(id, ctrl);
```

Comment:
I think you should construct an `std::unique_ptr` from `ctrl` explicitly or change the type of `ctrl` to `std::unique_ptr<Control>` to fix the build error.
BTW, could you explain how the ownership of controls is managed?

Reply:
Thanks for the hint, I added an explicit `unique_ptr`, it now works with the CI.

The ownership is token by `LinkedModelGroupView` when calling `addControl`. The doxygen comment said that already, but I added a further comment next to the new `unique_ptr`.